core/src/main/scala/org/apache/spark/sql/delta/files/TahoeCDCBaseFileIndex.scala

```diff
@@ -0,0 +1,104 @@
+
+abstract class TahoeCDCBaseFileIndex[T <: FileAction](spark: SparkSession,
+                                                  val filesByVersion: Seq[CDCDataSpec[T]],
+                                                  deltaLog: DeltaLog,
+                                                  path: Path,
+                                                  snapshot: Snapshot)
+  extends TahoeFileIndex(spark, deltaLog, path) {
+
+  def extractActionParameters(action: T):
+  (Map[String, String], Long, Long, Boolean, Map[String, String]) = {
+    action match {
+      case AddCDCFile(_, partitionValues, size, tags) => (partitionValues, size, 0, false, tags)
+      case r@RemoveFile(_, _, dataChange, extendedFileMetadata, partitionValues, size, tags) =>
+        if (!extendedFileMetadata.getOrElse(false)) {
+          // This shouldn't happen in user queries - the CDC flag was added at the same time as
+          // extended metadata, so all removes in a table with CDC enabled should have it. (The
+          // only exception is FSCK removes, which we screen out separately because they have
+          // dataChange set to false.)
+          throw DeltaErrors.removeFileCDCMissingExtendedMetadata(r.toString)
+        }
+        (partitionValues, size.getOrElse(0L), 0, dataChange, tags)
+      case AddFile(_, partitionValues, size, modificationTime, dataChange, _, tags) =>
+        (partitionValues, size, modificationTime, dataChange, tags)
+    }
+  }
+
+  override def tableVersion: Long = snapshot.version
+
+  override def matchingFiles(
+                              partitionFilters: Seq[Expression], dataFilters: Seq[Expression]):
+  Seq[AddFile] = {
+    val addFiles = filesByVersion
+      .flatMap {
+        case CDCDataSpec(version, timestamp, actions) =>
+          actions.map { action =>
+            val (partitionValues, size, modificationTime, dataChange, tags) =
+              extractActionParameters(action)
+
+            val newPartitionValues =
+              partitionValues ++
+                (Map(
+                  CDC_COMMIT_VERSION -> version.toString,
+                  CDC_COMMIT_TIMESTAMP -> Option(timestamp).map(_.toString).orNull)
+                  ++ additionalPartitionValues())
+
+            AddFile(
+              action.path, newPartitionValues, size, modificationTime, dataChange, tags = tags)
+          }
+      }
+    DeltaLog.filterFileList(
+      partitionSchema,
+      spark.createDataset(addFiles)(addFileEncoder).toDF(),
+      partitionFilters)
+      .as[AddFile](addFileEncoder)
+      .collect()
+  }
+
+  def additionalPartitionValues(): Map[String, String] = Map.empty
+
+  override def inputFiles: Array[String] = {
+    filesByVersion.flatMap(_.actions).map(f => absolutePath(f.path).toString).toArray
+  }
+
+  override def refresh(): Unit = {}
+
+  override def sizeInBytes: Long =
+    filesByVersion
+      .map(_.actions.map {
+        case a: AddFile => a.size
+        case r: RemoveFile => r.size.getOrElse(0L)
```

Comment:
can't this also be a `AddCDCFile`? e.g. in `TahoeChangeFileIndex` the type is `val filesByVersion: Seq[CDCDataSpec[AddCDCFile]]`.

how are existing tests passing?

Reply:
You are right, it can be `AddCDCFile`. I'll fix that.
Regarding tests -> It seems that none of the tests are actually triggering this match pattern. I'll add a test for that.